feat: Add trim_doctest_flag to google and numpy parsers

thatlittleboy opened the pull request:
Related to: https://github.com/mkdocstrings/mkdocstrings/issues/386
Porting the work done on the legacy handler (PR: https://github.com/mkdocstrings/pytkdocs/pull/134) over to griffe.

## Key PR pointers

* ❗️feat: I added in `example` as a section_kind, in addition to `examples` for both Google and Numpy. This is aligned with `pytkdocs` and also accepted by [Sphinx Napoleon](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstring-sections). Though I understand, the official Google/Numpy docs don't seem to be explicit about whether `example` is a valid title. Let me know if there was a conscious decision made to not include "example" in Griffe before --- I'll remove it from this PR.
* ❗️fix: I renamed the `keyword_parameters` in the `section_kind` dict; I presume this was a typo since I don't think keyword params is a thing.
* Feat: Implemented the functionality of removal of `# doctest:` and `<BLANKLINE>` via regex; in *both* cases where 1. line starts with `>>>` and 2. line is "in a code example block". This is activated via the new `trim_doctest_flags` parameter / option.
* As mentioned in the parent issue, I have set the default behaviour of `trim_doctest_flags` to be True. To keep consistent with Sphinx.
* Only implemented for Google and Numpy for now.
* Tests: Added test to check behaviour of `trim_doctest_flags` is as intended. (both when it is turned on, and turned off)
* Docs: Updated to reflect the new option. (Will raise a new issue for Sphinx missing `trim_doctest_flags` after this).


❗️About the implementation of the new parser option (and parser options **in general**). Because `trim_doctest_flags` option is only applicable to the examples section, the way I see it, there are 2 ways to go about the implementation.
Option 1 (naive way) is just adding simple if-else in the main `parse` while loop:
```python                                                                                                                               
def _read_examples_section(docstring, offset, trim_doctest_flags=True):                                                    
    # Add the `trim_doctest_flags` as a specific kwarg in this reader function.
    # All the other `_read_xxx` functions remain untouched. 
                                                                                                                                         
def parse(..., trim_doctest_flags=True, **options):                                                                                     
    # ...                                                                                                                                 
    if admonition_type.lower() == 'examples':                                                                                             
        sections, offset =  reader(docstring, offset + 1, trim_doctest_flags)
    else:
        sections, offset = reader(docstring, offset + 1)
 ```                                                                                                                                     
The alternative is what I am proposing in this PR. Make all `_read_xxx` functions accept arbitrary `options` kwargs.
And each reader function is responsible for extracting out the options that they need.

I think the latter is cleaner, especially if we need to extend to other options that might only apply in other sections but not examples, etc. Let me know if this is okay or if you have any better ideas.

thatlittleboy reviewed (commented):
Leaving some comments directly on the code lines, just to highlight some key points for you to review. 😄

pawamoy commented on src/griffe/docstrings/google.py:
```diff
@@ -32,13 +32,14 @@
     "parameters": DocstringSectionKind.parameters,
     "keyword args": DocstringSectionKind.other_parameters,
-    "keyword parameters": DocstringSectionKind.other_parameters,
+    "keyword arguments": DocstringSectionKind.other_parameters,
```
No, good catch, thanks :slightly_smiling_face:

pawamoy commented on src/griffe/docstrings/google.py:
```diff
@@ -32,13 +32,14 @@
     "exceptions": DocstringSectionKind.raises,
     "returns": DocstringSectionKind.returns,
     "yields": DocstringSectionKind.yields,
+    "example": DocstringSectionKind.examples,
```
I indeed consciously omitted `Example` because it is a common admonition type. I felt like only having `Examples` for special examples section was enough. So yeah, I'd like to keep it that way and you can remove this line :slightly_smiling_face:

pawamoy commented on src/griffe/docstrings/google.py:
```diff
@@ -402,8 +437,13 @@ def _read_receives_section(docstring: Docstring, offset: int) -> tuple[Docstring
+def _read_examples_section(  # noqa: WPS231
+    docstring: Docstring,
+    offset: int,
+    **options: Any,
+) -> tuple[DocstringSection | None, int]:
     text, new_offset = _read_block(docstring, offset)
+    trim_doctest_flags = options.get("trim_doctest_flags", True)
```
I guess it would be fine if you explicitly declare `trim_docstring_flags` in the signature while keeping `**options`. This would allow to both provide an annotation and a default without having to use `options.get` in the function's body.

pawamoy commented on src/griffe/docstrings/google.py:
```diff
@@ -497,18 +547,20 @@ def _is_empty_line(line) -> bool:
 
+    Additional parser options that are supported for Google style:
+        ignore_init_summary: Whether to ignore the summary in `__init__` methods' docstrings. Defaults to False.
+        trim_doctest_flags: Whether to remove doctest flags from Python example blocks. Defaults to True.
+
     Parameters:
         docstring: The docstring to parse.
-        ignore_init_summary: Whether to ignore the summary in `__init__` methods' docstrings.
-        **options: Additional parsing options.
+        **options: Additional parsing options. See description above.
```
As hinted above, I'm more in favor of explicitness, so declaring available options in the signature itself, while still keeping `**options` (even if unused). It makes for a better UX (autocompletion/suggestions in IDEs, documented type and default in API docs). Maybe when we have 15 different options we can reconsider :slightly_smiling_face:

pawamoy commented on src/griffe/docstrings/google.py:
```diff
@@ -519,8 +571,14 @@ def parse(  # noqa: WPS231
     in_code_block = False
     lines = docstring.lines
 
+    default_parser_options = {
+        "ignore_init_summary": False,
+        "trim_doctest_flags": True,
+    }
+    options = {**default_parser_options, **options}
```
Since I'd like to keep options in the signature, this can be changed to something like:

```python
options = {
    "ignore_init_summary": ignore_init_summary,
    "trim_doctest_flags": trim_doctest_flags,
}
```

pawamoy commented on src/griffe/docstrings/numpy.py:
```diff
@@ -71,6 +71,7 @@
     "receives": DocstringSectionKind.receives,
     "raises": DocstringSectionKind.raises,
     "warns": DocstringSectionKind.warns,
+    "example": DocstringSectionKind.examples,
```
There's no "admonition" feature with the Numpy parser (IIRC), but for the sake of consistency I believe we should drop this as well (also, maybe it'll support admonitions later).

pawamoy reviewed (changes requested):
Thanks for this high quality PR :slightly_smiling_face: 
A few answers and comments.

thatlittleboy commented:
It's late here so I'm calling it a day.
I'll try to address your review sometime during the week, but latest by the weekend if I end up being too busy with work ><

pawamoy reviewed (approved):
Thanks a lot!

pawamoy commented:
Do you want to keep your commits separated? If so, please squash the last two where they belong. Otherwise I can squash them all, as you prefer.

thatlittleboy commented:
I've rebased to remove adding the "Example" admonition, and adjusted the `feat: implement trim_doctest_flags` commit to incorporate the changes to options kwargs and use "Examples" in the tests rather than "Example".

pawamoy commented:
That's perfect, thanks again :rocket: :100: !

pawamoy commented:
Released in version 0.13.0 :smile:
